qcengine/programs/adcc.py

```diff
@@ -0,0 +1,155 @@
+            charge=mol.molecular_charge,
+            basis=model.basis,
+            #   conv_tol=model.scf_conv,
+            #   max_iter=,
+        )
+
+        # handle defaults nicely...
+        n_singlets = kws.pop("n_singlets", 3)
+
+        adcc.set_n_threads(config.ncores)
+        adcc_state = adcc.run_adc(scfres, method=model.method, n_singlets=n_singlets)
+
+        input_data = input_model.dict(encoding="json")
+        output_data = input_data.copy()
+        output_data["success"] = adcc_state.converged
+        provenance = Provenance(creator="adcc", version=self.get_version()).dict()
+        provenance["nthreads"] = adcc.get_n_threads()
+        output_data["provenance"] = provenance
+        output_data["properties"] = AtomicResultProperties()
+        output_data["return_result"] = adcc_state.excitation_energy
+
+        extract_props = input_model.driver == "properties"
+        output_data["extras"]["qcvars"] = self._extract_qcvars(adcc_state, extract_props)
+        return AtomicResult(**output_data)
+
+    def _extract_qcvars(self, state, extract_props=False):
+        qcvars = {}
+        name = state.method.name
+        NAME = name.upper()
+        is_cvs_adc3 = state.method.level >= 3 and state.ground_state.has_core_occupied_space
+        mp = state.ground_state
+        mp_energy = mp.energy(state.method.level if not is_cvs_adc3 else 2)
+        mp_corr = 0.0
+        qcvars[f"HF TOTAL ENERGY"] = mp.reference_state.energy_scf
+        if state.method.level > 1:
+            for level in range(2, state.method.level + 1):
+                if level >= 3 and is_cvs_adc3:
+                    continue
+                energy = mp.energy_correction(level)
+                mp_corr += energy
+                qcvars[f"MP{level} CORRELATION ENERGY"] = energy
+                qcvars[f"MP{level} TOTAL ENERGY"] = mp.energy(level)
+        qcvars["EXCITATION KIND"] = state.kind.upper()
+        qcvars[f"{NAME} ITERATIONS"] = state.n_iter
+        qcvars[NAME + " EXCITATION ENERGIES"] = state.excitation_energy
+        qcvars["NUMBER OF EXCITED STATES"] = len(state.excitation_energy)
+        if extract_props:
+            qcvars["HF DIPOLE"] = mp.reference_state.dipole_moment
+            if state.method.level > 1:
+                qcvars["MP2 DIPOLE"] = mp.dipole_moment(2)
+            # transition properties
+            qcvars[f"{NAME} TRANSITION DIPOLES (LEN)"] = state.transition_dipole_moment
+            qcvars[f"{NAME} TRANSITION DIPOLES (VEL)"] = state.transition_dipole_moment_velocity
+            qcvars[f"{NAME} OSCILLATOR STRENGTHS (LEN)"] = state.oscillator_strength
+            qcvars[f"{NAME} OSCILLATOR STRENGTHS (VEL)"] = state.oscillator_strength_velocity
+            qcvars[f"{NAME} ROTATIONAL STRENGTHS (VEL)"] = state.rotatory_strength
+            # state properties
+            qcvars[f"{NAME} STATE DIPOLES"] = state.state_dipole_moment
+        return qcvars
```

Comment:
That's duplicate code with psi4 if I remember correctly. Should we not just put this into adcc if that's needed from both places?

Reply:
The code is quite similar to what is done in Psi4, right.
In Psi4, it's mixed with some print-out...
Could be a function on `ExcitedStates`, e.g., `to_qcvars` or `to_dict`, and then inside Psi4 the variables would be set on the wave function.

Reply:
Yes that, but since it is pretty specific to this ecosystem perhaps a free function in just another file in `adcc`?

Reply:
I think we should finalize the routine here until we're happy with it and then transfer it adcc-side. Otherwise we probably make some mistake and have to create several adcc releases until it works fine 😉

Reply:
That sounds reasonable.